### ISSUE TITLE
Avoid linear scans through all users, streams, etc.

### ISSUE THREAD

**Issue by gnprice** [zulip/zulip-mobile#3339]
There are a number of places where we make a linear search through an array of potentially thousands of items, when we really just want a specific item with a known ID or name.

This is a pattern that can be devastating for performance -- in particular, as soon as it happens in a loop, we have quadratic behavior. This is a common class of performance bug found in lots of software:
https://accidentallyquadratic.tumblr.com/post/113840433022/why-accidentally-quadratic
and we should avoid it.

Here's a rough catalog of where we do this:
```
$ git grep -ohP '\w+\.find\(.*?\)' src/ | sort | grep --color '^\w*'
actionSheetButtons.find(x => x.title === title)
allStreams.find(stream => stream.stream_id === subscription.stream_id)
allUsers.find(u => u.email === narrow[0].operand)
allUsers.find(x => x.email === email)
allUsers.find(x => x.email === r)
handlers.find(x => x.isFunc(narrow)
messages.find(msg => !isMessageRead(msg, flags, subscriptions, mute)
messages.find(x => x.id === messageId)
messages.find(x => x.id === messageId)
messages.find(x => x.isFunc(narrow)
mute.find(x => x[0] === stream.name && x[1] === name)
selected.find(user => user.user_id === item.user_id)
selected.find(x => x.email === email)
state.find(item => item.timestamp === action.outbox.timestamp)
state.find(y => x.stream_id === y.stream_id)
state.find(y => x.stream_id === y.stream_id)
streams.find(s => s.name === narrow[0].operand)
streams.find(s => s.name === narrow[0].operand)
streams.find(s => s.name === narrow[0].operand)
streams.find(s => s.name === narrow[0].operand)
streams.find(s => s.name === narrow[0].operand)
streams.find(stream_ => narrow[0].operand === stream_.name)
streams.find(sub => narrow[0].operand === sub.name)
streams.find(x => x.name === narrow[0].operand)
streams.find(x => x.name === narrow[0].operand)
streams.find(x => x.name === narrow[0].operand)
streams.find(x => x.name === narrow[0].operand)
streams.find(x => x.name === narrow[0].operand)
streams.find(x => x.stream_id === streamId)
streams.find(y => x && x.stream_id === y.stream_id)
subscriptions.find(sub => narrow[0].operand === sub.name)
subscriptions.find(sub => streamName === sub.name)
subscriptions.find(x => x.name === display_recipient)
subscriptions.find(x => x.name === message.display_recipient)
subscriptions.find(x => x.name === message.display_recipient)
subscriptions.find(x => x.name === message.display_recipient)
subscriptions.find(x => x.name === message.display_recipient)
subscriptions.find(x => x.name === narrow[0].operand)
subscriptions.find(x => x.stream_id === streamId)
subscriptions.find(y => x && y && x.stream_id === y.stream_id)
unreadHuddles.find(x => x.user_ids_string === userIds)
unreadPms.find(x => x.sender_id === sender.user_id)
users.find(user => user.email === email)
users.find(user => user.email === narrow[0].operand)
users.find(user => user.email === userEmail)
users.find(user => user.user_id === userId)
users.find(u => u.email === narrow[0].operand)
users.find(x => x.email === email)
```

A few of those are innocuous, searching through a fixed list of a handful of entries. But the majority of them appear to be searching through our entire list of known users, or non-bot users, or streams, or subscriptions -- all of them lists of unbounded length -- just looking for an ID, email, or name.

Instead, we should make these lookups through efficient, indexed data structures. A JS object works great. A `Map` works equally great, and can be a bit cleaner in the code.

Of these, the searches through users are the highest priority to fix; that can be a very long list indeed, and I suspect these are already a cause of sluggishness for users of large orgs like chat.zulip.org.

**Comment by gnprice** [comment]
I just pushed a series of commits that fix some of these. See the two commits shown above that link to this issue; plus 853bb9508 which sets up one of them by adding a cached selector to construct the appropriate data structure.

If you rerun the search in the OP, you'll find there are still 3 cases of `users.find`, plus all of the `streams.find` and others. So, plenty more to be done!

**Comment by jainkuniya** [comment]
I can pick up the rest one.
@gnprice Please assign me this.

@zulipbot claim

**Comment by borisyankov** [comment]
That is not entirely correct, because:
* `O(n^2)` might be terrible in theory, but when `n` is a small number and when the operation is fast, then this is not relevant
* often it is not quadratic but `O(n * m)` where `m` is a small number thus not as bad
* only `Users` and maybe `Messages` would be reaching values that it is a problem (Users say 15k * 15k = 225mil obviously)

Selectors are not a universal solution when we are talking about frequently changing data as they get recalculated. For Users that is clearly not a problem as they change rarely.

For Streams and Subscriptions, this would be a 'premature optimization'

**Comment by gnprice** [comment]
Thanks @jainkuniya -- assigned! :slightly_smiling_face:

**Comment by gnprice** [comment]
@borisyankov Indeed, the cases I'm most concerned about are where we search through the list of users, which is the longest of these.

We routinely need to get details about users for things like showing their avatars in a message list, where we might do it for hundreds of messages at once, or for rendering a user picker. Those are a large `m`. And `n * m` is very much a quadratic function. :stuck_out_tongue: 

> Selectors are not a universal solution when we are talking about frequently changing data as they get recalculated. For Users that is clearly not a problem as they change rarely.

Fortunately the same thing is even more true of streams and subscriptions.

---

As for recalculating the maps -- clearly for a complete solution we want to be maintaining our data structures in an efficient way, in addition to looking things up in them efficiently. The way we accumulate our arrays in the Redux state is actually already quadratic in itself. Here's a quick demo with `node`:
```
$ for n in 1000 2000 4000 8000 16000; do node -e 'const n = '"$n"'; let a = []; const start = Date.now(); for (let i = 0; i < n; ++i) { a = [...a, i]; }; console.log(`${n}: ${(Date.now() - start)}ms`)'; done
1000: 5ms
2000: 11ms
4000: 32ms
8000: 119ms
16000: 499ms
```

And yikes, it's 30-40x worse than that when maintaining an `Object` instead:
```
$ for n in 1000 2000 4000 8000 16000; do node -e 'const n = '"$n"'; let o = {}; const start = Date.now(); for (let i = 0; i < n; ++i) { o = {...o, [i]: i}; }; console.log(`${n}: ${(Date.now() - start)}ms`)'; done
1000: 86ms
2000: 335ms
4000: 1320ms
8000: 5381ms
16000: 21818ms
```

Interestingly, maintaining a `Map` in the same (inefficient, quadratic) way is "only" 12-15x slower than `Array`, so over twice as fast as `Object`:
```
$ for n in 1000 2000 4000 8000 16000; do node -e 'const n = '"$n"'; let m = new Map(); const start = Date.now(); for (let i = 0; i < n; ++i) { m = new Map(m.entries()); m.set(i, i); }; console.log(`${n}: ${(Date.now() - start)}ms`)'; done
1000: 26ms
2000: 98ms
4000: 393ms
8000: 1653ms
16000: 7005ms
```

Still, none of those are great.

Efficiently maintaining data structures without mutating them is a bit subtle, but there are well-understood techniques for doing so (called "persistent data structures"). One implementation of those in JS is [Immutable.js](https://github.com/facebook/immutable-js/). Doing the same quick demo:
```
$ for n in 1000 2000 4000 8000 16000; do node -e 'const n = '"$n"'; const { Map: ImMap } = require("immutable"); let m = ImMap(); const start = Date.now(); for (let i = 0; i < n; ++i) { m = m.set(i, i); }; console.log(`${n}: ${(Date.now() - start)}ms`)'; done
1000: 5ms
2000: 8ms
4000: 12ms
8000: 16ms
16000: 22ms
```

That is a lot better!

**Comment by gnprice** [comment]
An update: we're now well on our way to solving this class of problem by maintaining efficient data structures in the first place, with Immutable.js. See #3949 for detailed progress.
